### PR TITLE
Optimizer/adam3

### PR DESCRIPTION
**Pull request by KOHbDS**
Test `Adam` step by step.

### CODE REVIEW

File: client/tinychain/ml/dnn.py [hunk 1]
```diff
@@ -87,12 +99,15 @@ def backward(self, inputs, loss):
...
+                    param_list.extend(layer_param_list)
 
-                return loss, layer_gradients
+                return param_list
+            
+            def get_param_list(self)  -> List[Parameter]:
+                return functools.reduce(operator.add, [layer.get_param_list() for layer in layers], [])
 
-        return DNN(layers)
+        return DNN(layers)
```
Comment (haydnv) [inline client/tinychain/ml/dnn.py:113]: [nit] add a newline at the end of the file

File: tests/client/test_adam.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+  
+import matplotlib.pyplot as plt
+import networkx as nx
+import tinychain as tc
+from tinychain.collection.tensor import Tensor
+
+# np.random.seed(42)
+# print(np.random.rand(1))
+
+# TODO
```
Comment (haydnv) [inline tests/client/test_adam.py:18]: [nit] TODO: what?

**Review by haydnv — CHANGES_REQUESTED** [review]
Looks good! Please remove the commented-out code and revert the `.gitignore` and `Cargo.toml` files.

Also, you're going to get some more merge conflicts from commit 87bb678. Please rebase or merge to get the updates to `decorators.closure` and `ml.optimizer.train`, or feel free to cherry-pick changes from this branch into a new branch.

**Comment by haydnv** [comment]
Please revert the `Cargo.toml` file and remove the `local.Dockerfile` and `pycache` files from Git

**Comment by haydnv** [comment]
I ran `testAdam_a` in `test_adam.py` several times and it really looks correct to me. Here's how I compared the initial and final state of the `DNN`:

```python
    cxt.nn = tc.ml.dnn.DNN.load([...])

    tc.print_json(HOST.post(ENDPOINT, cxt))  # record the initial state

    cxt.optimizer = tc.ml.optimizer.Adam.create(param_list=cxt.nn.get_param_list())
    cxt.result = tc.After(tc.ml.optimizer.train(...), cxt.nn)
    tc.print_json(HOST.post(ENDPOINT, cxt))  # get the final state to compare
```

It's true that the trainable parameters don't change very much but that's just because of the learning rate and the number of iterations. I think you should just clean up the code a little so the tests in `test_neural_net.py` run using Adam optimization and the `train` method signature matches the one in the `main` branch: `def train(model, optimizer, inputs, labels, cost, train_while)` and then this will be ready to merge.

**Review by haydnv — CHANGES_REQUESTED** [review]

File: client/tinychain/ml/optimizer.py [hunk 3]
```diff
@@ -1,58 +1,85 @@
...
+from tinychain.ml import DiffedParameter, Parameter
 from tinychain.ref import After, While
 from tinychain.state import Map
 from tinychain.value import F32, UInt
 
 
 class Optimizer(Map):
     @abstractmethod
-    def optimize(self, i, gradient, inputs, loss):
+    def optimize(self):
```
Comment (haydnv) [inline client/tinychain/ml/optimizer.py:14]: [nit] it would be nice to include the `param_list` arg here

File: client/tinychain/ml/optimizer.py [hunk 4]
```diff
@@ -1,58 +1,85 @@
...
-        _loss, deltas = gradient.backward(inputs, loss)
-        return update(gradient, deltas)
+    @classmethod
+    def create(cls, param_list: t.List[Parameter],  beta1=F32(0.9), beta2=F32(0.999), lr=F32(1e-3), eps=F32(1e-8)):
+        """Create a new `Adam` optimizer with the given hyperparameters.
+        'lr': learning rate;
+        'beta1': The exponential decay rate for the 1st moment estimates;
+        'beta2': The exponential decay rate for the 2nd moment estimates;
+        'eps': A small constant for numerical stability.
+        Args:
```
Comment (haydnv) [inline client/tinychain/ml/optimizer.py:44]: "Args:" should be on line 40, before the argument list

File: host/tensor/Cargo.toml [hunk 5]
```diff
@@ -12,7 +12,7 @@ keywords = ["tinychain", "tensor", "gpu", "ml", "parallel"]
 categories = ["concurrency", "data-structures", "hardware-support", "mathematics"]
 
 [dependencies]
-afarray = "~0.12.1"
+afarray = "0.12"
```
Comment (haydnv) [inline host/tensor/Cargo.toml:15]: please revert this file and remove `__pycache__` from Git

File: tests/client/test_neural_net.py [hunk 6]
```diff
@@ -30,93 +30,111 @@ def truncated_normal(size, mean=0., std=None):
...
+            cxt.optimizer = tc.ml.optimizer.GradientDescent.create(lr=LEARNING_RATE)
+        elif optim == 'Adam':
+            cxt.optimizer = tc.ml.optimizer.Adam.create(param_list=cxt.nn.get_param_list(), lr=LEARNING_RATE)
         cxt.result = tc.ml.optimizer.train(cxt.nn, cxt.optimizer, cxt.inputs, cxt.labels, cost, train_while)
 
         response = self.host.post(ENDPOINT, cxt)
+
         self.assertLess(response["i"], MAX_ITERATIONS, "failed to converge")
 
+    def testAdam(self):
```
Comment (haydnv) [inline tests/client/test_neural_net.py:122]: this can be renamed `testXor_3layer`

File: tests/client/test_neural_net.py [hunk 7]
```diff
@@ -30,93 +30,111 @@ def truncated_normal(size, mean=0., std=None):
...
+        cxt.output_layer = self.create_layer('layer1', 2, 1, tc.ml.Sigmoid())
+        cxt.nn = tc.ml.dnn.DNN.load([cxt.input_layer0, cxt.output_layer])
 
-        self.execute(cxt)
+        self.execute(cxt, optim='GrDes')
 
-    def execute(self, cxt):
-        def cost(output, labels):
-            return (output - labels)**2
+    def execute(self, cxt, optim):
```
Comment (haydnv) [inline tests/client/test_neural_net.py:101]: you can remove the `optim` arg and just always use Adam optimizer

File: tests/client/test_neural_net.py [hunk 8]
```diff
@@ -30,93 +30,111 @@ def truncated_normal(size, mean=0., std=None):
...
+        cxt.labels = load(labels)
 
-        cxt.input_layer = self.create_layer(2, 1, tc.ml.ReLU())
+        cxt.input_layer = self.create_layer('layer0', 2, 1, tc.ml.ReLU())
         cxt.nn = tc.ml.dnn.DNN.load([cxt.input_layer])
 
-        self.execute(cxt)
+        self.execute(cxt, optim='GrDes')
 
     def testXor(self):
```
Comment (haydnv) [inline tests/client/test_neural_net.py:86]: this can be renamed `testXor_2layer`

**Review by haydnv — APPROVED** [review]
Looks good! I'm going to look into the issue that you raised about slow training times but I don't think we should block this PR on that

**Review by haydnv — APPROVED** [review]
Looks good! Please just delete the `tests/client/__pycache__directory` though
